feat: Add ipprefix cast operators for ipaddr [2/n]

yuandagits opened the pull request:
Summary:
Support casting ipprefix -> ipaddress, and ipaddress -> ipprefix.

We just need one new helper function, tryIpPrefixLengthFromIPAddressType, which takes in int128_t, gets the ippaddress, and then we check if the ippaddress is ipv4 or ipv6 mapped. We return 32 for ipv4 and 128 for ipv6 mapped.

Differential Revision: D65641994

mohsaka commented on velox/functions/prestosql/types/IPPrefixType.cpp:
```diff
@@ -120,6 +148,33 @@ class IPPrefixCastOperator : public exec::CastOperator {
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      auto intIpAddr = ipAddrVector->valueAt(row);
+      const auto tryPrefixLength =
```
I don't think this is particularly necessary as the entire range of the 16 byte array are all valid ips.

yuandagits replied:
Hm make sense, but I still need to get the prefix length. I don't think the cost of the check is expensive

mohsaka commented on velox/functions/prestosql/types/IPPrefixType.cpp:
```diff
@@ -120,6 +148,33 @@ class IPPrefixCastOperator : public exec::CastOperator {
+        return;
+      }
+
+      auto writer = exec::VectorWriter<Row<int128_t, int8_t>>();
```
Is the writer faster than setting the vector directly? Not sure.

mohsaka commented on velox/functions/prestosql/types/IPPrefixType.cpp:
```diff
@@ -78,14 +94,26 @@ class IPPrefixCastOperator : public exec::CastOperator {
   }
 
  private:
+  static void castToIpAddress(
```
To cast functions aren't needed and can become a maintenance issue since they are not called.

mohsaka commented on velox/functions/prestosql/types/IPPrefixType.cpp:
```diff
@@ -70,6 +80,12 @@ class IPPrefixCastOperator : public exec::CastOperator {
         return castToString(input, context, rows, *result);
+      case TypeKind::HUGEINT: {
+        if (isIPAddressType(input.type())) {
+          return castToIpAddress(input, context, rows, *result);
```
Can return an Fail. Something like this
```
VELOX_FAIL("We should be using IPPREFIX castTo function");
```

Or can just do nothing. It should never be called anyways since the castFrom calls come first.

yuandagits replied:
Ah good catch, will update

mohsaka reviewed (changes requested):
You can take a look at my updated PRs with addressed comments.

https://github.com/facebookincubator/velox/pull/11309/files
https://github.com/facebookincubator/velox/pull/11378/files

Also here is the functions one if you are doing that too
https://github.com/facebookincubator/velox/pull/11407

Will be going on vacation so feel free to get yours merged!👍

yuandagits commented:
> You can take a look at my updated PRs with addressed comments.
> 
> https://github.com/facebookincubator/velox/pull/11309/files https://github.com/facebookincubator/velox/pull/11378/files
> 
> Also here is the functions one if you are doing that too #11407
> 
> Will be going on vacation so feel free to get yours merged!👍

thanks! I wasn't aware of your PRs 😅, I will take a look at your pointers

Yuhta commented on velox/functions/prestosql/types/IPPrefixType.cpp:
```diff
@@ -120,6 +128,30 @@ class IPPrefixCastOperator : public exec::CastOperator {
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* rowVectorResult = result.as<RowVector>();
+    const auto* ipAddrVector = input.as<SimpleVector<int128_t>>();
```
Use `DecodedVector` is both faster and safer

Yuhta commented on velox/functions/prestosql/types/IPPrefixType.cpp:
```diff
@@ -120,6 +128,30 @@ class IPPrefixCastOperator : public exec::CastOperator {
+        return;
+      }
+
+      rowVectorResult->childAt(0)->as<FlatVector<int128_t>>()->set(
```
Move type casts out of the per-row loop.  You can also use `asChecked` once it's outside loop.
